### ISSUE TITLE
std::out_of_range issue

### ISSUE THREAD

**Issue by Nimisha-Mathews** [p4lang/behavioral-model#306]
Hi
I have a 3 switch circular topology, with spanning tree configured. I noticed that intermittently I get the following issue on one or more of the switches. All the values passed to the switch_apis are ok.

This is the error:

**_terminate called after throwing an instance of 'std::out_of_range'
  what(): _Map_base::at_**

after which my process terminates.

Attached gdb to the process and got the following backtrace. Could you please help me figure out what is going wrong.

(gdb) bt
#0  0x00007f0f4c4e1248 in __GI_raise (sig=sig@entry=6) at /usr/src/debug/glibc/2.23-r0/git/sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007f0f4c4e269a in __GI_abort () at /usr/src/debug/glibc/2.23-r0/git/stdlib/abort.c:89
#2  0x00007f0f454d746d in __gnu_cxx::__verbose_terminate_handler() () from target:/usr/lib/libstdc++.so.6
#3  0x00007f0f454d54d6 in ?? () from target:/usr/lib/libstdc++.so.6
#4  0x00007f0f454d5521 in std::terminate() () from target:/usr/lib/libstdc++.so.6
#5  0x00007f0f454d5738 in __cxa_throw () from target:/usr/lib/libstdc++.so.6
#6  0x00007f0f498b04d1 in unsigned int apache::thrift::transport::readAll<apache::thrift::transport::TBufferBase>(apache::thrift::transport::TBufferBase&, unsigned char*, unsigned int) ()
   from target:/usr/lib/openvswitch/plugins/libovs_p4_sim_plugin.so
#7  0x00007f0f498b139d in apache::thrift::protocol::TBinaryProtocolT<apache::thrift::transport::TTransport, apache::thrift::protocol::TNetworkBigEndian>::readMessageBegin(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, apache::thrift::protocol::TMessageType&, int&) () from target:/usr/lib/openvswitch/plugins/libovs_p4_sim_plugin.so
#8  0x00007f0f46412ad0 in bm_runtime::standard::StandardClient::recv_bm_mt_modify_entry() () from target:/usr/lib/libruntimestubs.so.0
#9  0x00007f0f4921a624 in p4_pd_dc_spanning_tree_table_modify_with_set_stp_state () from target:/usr/lib/libpd.so.0
#10 0x00007f0f498c8a04 in switch_pd_spanning_tree_table_update_entry () from target:/usr/lib/openvswitch/plugins/libovs_p4_sim_plugin.so
#11 0x00007f0f498d5c9d in switch_api_stp_port_state_set () from target:/usr/lib/openvswitch/plugins/libovs_p4_sim_plugin.so

Thanks for any pointers!

**Comment by antoninbas** [comment]
This is not really an helpful trace. You give a gdb trace for the driver process, but the crash in the driver process is probably caused by the `std::out_of_range` exception in the bmv2 model process. We need the gdb trace for the bmv2 process.

**Comment by Nimisha-Mathews** [comment]
[nano.txt](https://github.com/p4lang/behavioral-model/files/834761/nano.txt)

Attached are the pipeline logs that were saved on one of the switches when this issue occurred. Would this be helpful?

**Comment by antoninbas** [comment]
That doesn't help. These are just packet logs and don't tell us where in the bmv2 C++ code the exception happened.

**Comment by Nimisha-Mathews** [comment]
How to enable debug symbols on bmv2? I am able to capture a gdb stack trace on simple switch, but without debug symbols it does not look meaningful.

**Comment by antoninbas** [comment]
I think symbols are enabled by default by autoconf. `-O2` is also the default, which can make debugging with gdb a little harder.
If you believe that you don't have symbols, then I recommend that you recompile & reinstall simple_switch. You will need to run configure again: `./configure 'CXXFLAGS=-g -O0' 'CFLAGS=-g -O0'`. If you are using an existing build directory, don't forget to run `make clean` or the new flags won't be taken into account.

**Comment by Nimisha-Mathews** [comment]
[switch_bmv2.json.txt](https://github.com/p4lang/behavioral-model/files/846280/switch_bmv2.json.txt)

Attached is the json file generated by the p4 compiler. I am not able to reproduce the issue all the time, it happens intermittently.
I was not able to get a simple_switch trace.
The following is the switch trace I got.

#0  0x00007f1a79183248 in __GI_raise (sig=sig@entry=6) at /usr/src/debug/glibc/2.23-r0/git/sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007f1a7918469a in __GI_abort () at /usr/src/debug/glibc/2.23-r0/git/stdlib/abort.c:89
#2  0x00007f1a720ec46d in __gnu_cxx::__verbose_terminate_handler() () from target:/usr/lib/libstdc++.so.6
#3  0x00007f1a720ea4d6 in ?? () from target:/usr/lib/libstdc++.so.6
#4  0x00007f1a720ea521 in std::terminate() () from target:/usr/lib/libstdc++.so.6
#5  0x00007f1a720ea738 in __cxa_throw () from target:/usr/lib/libstdc++.so.6
#6  0x00007f1a764bfc07 in apache::thrift::transport::readAll<apache::thrift::transport::TBufferBase> (trans=..., buf=0x7fff433ed990 " \332>C\377\177", len=4)
    at /usr/include/thrift/transport/TTransport.h:43
#7  0x00007f1a764c01e8 in apache::thrift::transport::TTransport::readAll (len=4, buf=0x7fff433ed990 " \332>C\377\177", this=<optimized out>)
    at /usr/include/thrift/transport/TTransport.h:121
#8  apache::thrift::protocol::TBinaryProtocolT<apache::thrift::transport::TTransport, apache::thrift::protocol::TNetworkBigEndian>::readI32 (
    i32=<synthetic pointer>, this=0x1291030) at /usr/include/thrift/protocol/TBinaryProtocol.tcc:371
#9  apache::thrift::protocol::TBinaryProtocolT<apache::thrift::transport::TTransport, apache::thrift::protocol::TNetworkBigEndian>::readMessageBegin (
    this=0x1291030, name=..., messageType=@0x7fff433eda1c: 32538, seqid=@0x7fff433eda18: 0) at /usr/include/thrift/protocol/TBinaryProtocol.tcc:203
#10 0x00007f1a764c03ed in apache::thrift::protocol::TVirtualProtocol<apache::thrift::protocol::TBinaryProtocolT<apache::thrift::transport::TTransport, apache::thrift::protocol::TNetworkBigEndian>, apache::thrift::protocol::TProtocolDefaults>::readMessageBegin_virt (this=<optimized out>, name=...,
    messageType=<optimized out>, seqid=<optimized out>) at /usr/include/thrift/protocol/TVirtualProtocol.h:403
#11 0x00007f1a73027ad0 in bm_runtime::standard::StandardClient::recv_bm_mt_modify_entry() () from target:/usr/lib/libruntimestubs.so.0

**Comment by antoninbas** [comment]
I looked at the code and IMO the only things that would explain this are:
- the table name provided for the modify_entry is invalid or
- the action name provided is not valid for this table
However I don't see how this could happen here. The PD function `p4_pd_dc_spanning_tree_table_modify_with_set_stp_state` should call modify entry with table name "spanning_tree" and action name "set_stp_state", both of which are correct based on the JSON input you provided.
Without knowing exactly where the exception was thrown in simple_switch, I cannot really spend more time investigating this.

**Comment by Nimisha-Mathews** [comment]
This is the simple_switch backtrace

(gdb) bt
#0  0x00007f01ae655248 in __GI_raise (sig=sig@entry=6) at /usr/src/debug/glibc/2.23-r0/git/sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007f01ae65669a in __GI_abort () at /usr/src/debug/glibc/2.23-r0/git/stdlib/abort.c:89
#2  0x00007f01afcf846d in __gnu_cxx::__verbose_terminate_handler() () from target:/usr/lib/libstdc++.so.6
#3  0x00007f01afcf64d6 in ?? () from target:/usr/lib/libstdc++.so.6
#4  0x00007f01afcf6521 in std::terminate() () from target:/usr/lib/libstdc++.so.6
#5  0x00007f01afd20368 in ?? () from target:/usr/lib/libstdc++.so.6
#6  0x00007f01ae9ca394 in start_thread (arg=0x7f01aa097700) at /usr/src/debug/glibc/2.23-r0/git/nptl/pthread_create.c:333
#7  0x00007f01ae70973d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109

**Comment by antoninbas** [comment]
Hi,
Thanks a lot for all the information you have been providing. The exception is being thrown in a thread and the stack is unwound before calling terminate, which is why we don't see anything useful in the backtrace.
If you have some more patience :), I pushed some code to a branch (https://github.com/p4lang/behavioral-model/tree/nimisha-except), which should help use locate the exception. If you don't mind, I would really appreciate if you could try to reproduce the exception using that code. It should print some information to stdout which will help me track down the issue.

**Comment by Nimisha-Mathews** [comment]
I pulled code from the branch and tried it out, I was able to reproduce the issue. However, I did not obtain any of the prints. I will need to check if the prints get redirected to some other place, and will get back to you.

**Comment by Nimisha-Mathews** [comment]
None of the print statements are coming up. 
I verified that in general cases they come up on screen, but none of the Exception handling prints you have put are being printed. Maybe the expected exceptions are not caught?

**Comment by antoninbas** [comment]
What do you mean by "in general cases they come up on screen"?
Did the backtrace change at all? It should, given that I added a `noexcept` qualifier to `bm_mt_modify_entry`...
I don't see how nothing could be printed :(. I'm pretty sure I covered everything relevant with catch statements. Of course I was assuming that the issue was somewhere in modify_entry and that may have been an invalid assumption.

**Comment by Nimisha-Mathews** [comment]
By "general cases" I meant that I added some other std::cout prints in the code, but not in the try/catch block, and I saw them print out.
Unfortunately, the back trace remains the same too. 

A while ago, I had got a switch crash, and I had got a different back trace, pasted below. However, I was not able to reproduce this issue again. (This is not a simple_switch trace).

(gdb) bt
#0  0x00007fbaf372f248 in __GI_raise (sig=sig@entry=6) at /usr/src/debug/glibc/2.23-r0/git/sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007fbaf373069a in __GI_abort () at /usr/src/debug/glibc/2.23-r0/git/stdlib/abort.c:89
#2  0x00007fbaec6d746d in __gnu_cxx::__verbose_terminate_handler() () from target:/usr/lib/libstdc++.so.6
#3  0x00007fbaec6d54d6 in ?? () from target:/usr/lib/libstdc++.so.6
#4  0x00007fbaec6d5521 in std::terminate() () from target:/usr/lib/libstdc++.so.6
#5  0x00007fbaec6d5738 in __cxa_throw () from target:/usr/lib/libstdc++.so.6
#6  0x00007fbaf0ab04d1 in unsigned int apache::thrift::transport::readAll<apache::thrift::transport::TBufferBase>(apache::thrift::transport::TBufferBase&, unsigned char*, unsigned int) ()
   from target:/usr/lib/openvswitch/plugins/libovs_p4_sim_plugin.so
#7  0x00007fbaf0ab139d in apache::thrift::protocol::TBinaryProtocolT<apache::thrift::transport::TTransport, apache::thrift::protocol::TNetworkBigEndian>::readMessageBegin(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, apache::thrift::protocol::TMessageType&, int&) () from target:/usr/lib/openvswitch/plugins/libovs_p4_sim_plugin.so
#8  0x00007fbaed612400 in bm_runtime::standard::StandardClient::recv_bm_mt_delete_entry() () from target:/usr/lib/libruntimestubs.so.0
#9  0x00007fbaf0400b0d in p4_pd_dc_smac_table_delete () from target:/usr/lib/libpd.so.0

This does not look like it is taking the same path of "modify_entry", but rather it is a delete op. I was wondering if this would give any useful hint. I did not mention it earlier, because I have hit this issue only once.

**Comment by Nimisha-Mathews** [comment]
Hi
I tried "catch throw" on gdb. And I am getting the following backtrace for simple_switch.

(gdb) bt
#0  __cxxabiv1::__cxa_throw (obj=obj@entry=0x7f685a941540, tinfo=0x7f686f32eaf8 <typeinfo for std::out_of_range>, dest=0x7f686f063180 <std::out_of_range::~out_of_range()>)
    at ../../../../../../../../../work-shared/gcc-5.3.0-r0/gcc-5.3.0/libstdc++-v3/libsupc++/eh_throw.cc:62
#1  0x00007f686f075d27 in std::__throw_out_of_range (__s=__s@entry=0x51e273 "_Map_base::at")
    at ../../../../../../../../../../work-shared/gcc-5.3.0-r0/gcc-5.3.0/libstdc++-v3/src/c++11/functexcept.cc:90
#2  0x00000000004e10cf in std::__detail::_Map_base<unsigned long, std::pair<unsigned long const, bm::McSimplePre::L1Entry>, std::allocator<std::pair<unsigned long const, bm::McSimplePre::L1Entry> >, std::__detail::_Select1st, std::equal_to<unsigned long>, std::hash<unsigned long>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true>, true>::at (__k=@0x7f68693ec110: 0, this=0x149bb88) at /usr/include/c++/5.3.0/bits/hashtable_policy.h:665
#3  std::unordered_map<unsigned long, bm::McSimplePre::L1Entry, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<std::pair<unsigned long const, bm::McSimplePre::L1Entry> > >::at (__k=@0x7f68693ec110: 0, this=0x149bb88) at /usr/include/c++/5.3.0/bits/unordered_map.h:689
#4  bm::McSimplePreLAG::replicate (this=0x149bb50, ingress_info=...) at /usr/src/debug/ops-p4dp/gitAUTOINC+69bb29d1a961a3c1bc87-r0/git/src/bm_sim/simple_pre_lag.cpp:163
#5  0x00000000004560ca in SimpleSwitch::ingress_thread (this=0x1497580) at /usr/src/debug/ops-p4dp/gitAUTOINC+69bb29d1a961a3c1bc87-r0/git/targets/simple_switch/simple_switch.cpp:390
#6  0x00007f686f0782b0 in std::(anonymous namespace)::execute_native_thread_routine (__p=<optimized out>)
    at ../../../../../../../../../../work-shared/gcc-5.3.0-r0/gcc-5.3.0/libstdc++-v3/src/c++11/thread.cc:84
#7  0x00007f686dd22394 in start_thread (arg=0x7f68693ef700) at /usr/src/debug/glibc/2.23-r0/git/nptl/pthread_create.c:333
#8  0x00007f686da6173d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109

**Comment by antoninbas** [comment]
Thanks this is very useful. Looks like a multicast L1 node has been destroyed without the multicast group being updated. I did some cleanup of the multicast code that hopefully will solve the issue once it is merged into master (see #329).

**Comment by Nimisha-Mathews** [comment]
Thanks very much for all your help! This review fixed the issue I was seeing.

**Comment by antoninbas** [comment]
Thank you for being so patient and providing all this information.
